internal/proxyservice/http/proxy_service.go

```diff
@@ -12,18 +19,269 @@ import (
+	return &subservice
+}
+
+
+// ServeHTTP exists to implement the go_http.Handler interface
+func (proxy *proxyService) ServeHTTP(w gohttp.ResponseWriter, r *gohttp.Request) {
+	logger := proxy.logger
+
+	// Log request
+
+	logMsg := "Got request %v %v %v %v"
+	logger.Infoln(logMsg, r.URL.Path, r.Host, r.Method, r.URL.String())
+
+	// Validate request
+
+	if !proxy.validateProxyServerRules(w, r) {
+		return
+	}
+
+	// Remove headers intended only for proxy server.
+
+	r.Header.Del("Proxy-Connection")
+	r.Header.Del("Proxy-Authenticate")
+	r.Header.Del("Proxy-Authorization")
+
+	// Select a subservice
+
+	subservice := proxy.selectSubservice(r)
+
+	// No match: Send request with no authentication.
+	if subservice == nil {
+		proxy.handleRequest(w, r)
+		return
+	}
+
+	// Get current credential values
+
+	creds, err := subservice.RetrieveCredentials()
+	defer internal.ZeroizeCredentials(creds)
+
+	if err != nil {
+		gohttp.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Authenticate request
+
+	err = subservice.Authenticate(r, creds)
+	if err != nil {
+		gohttp.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Send request to target service
+
+	proxy.handleRequest(w, r)
+}
+
+// validateProxyServerRules ensures that the request being made is a valid
+// request for a proxy server.
+func (proxy *proxyService) validateProxyServerRules(
+	w gohttp.ResponseWriter, r *gohttp.Request,
+) bool {
+
+	if r.Method == "CONNECT" {
+		gohttp.Error(w, "CONNECT is not supported.", 405)
+		return false
+	}
+
+	if !r.URL.IsAbs() {
+		errMsg := "This is a proxy server. Non-proxy requests aren't allowed."
+		gohttp.Error(w, errMsg, 500)
+		return false
+	}
+
+	return true
+}
+
+// handleRequest sends the request to the target service and writes the response
+// back to the client.
+func (proxy *proxyService) handleRequest(
+	w gohttp.ResponseWriter, r *gohttp.Request,
+) {
+	logger := proxy.logger
+
+	// TODO: Must be reset for... some reason?
+	r.RequestURI = ""
+
+	// Send request to target service
+
+	resp, err := proxy.transport.RoundTrip(r)
+	if err != nil {
+		logger.Debugf("Error: %v\n", err)
+		gohttp.Error(w, err.Error(), 503)
+		return
+	}
+
+	// Send response to client (everything below)
+
+	logger.Debugf("Received response status: %s\n", resp.Status)
+
+	copyHeaders(w.Header(), resp.Header)
+
+	w.WriteHeader(resp.StatusCode)
+
+	// TODO: Is this the right thing to do with this error?
+	_, err = io.Copy(w, resp.Body)
+	if err != nil {
+		logger.Errorf("Can't write response to body: %s\n", err)
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		logger.Debugf("Can't close response body %v\n", err)
+	}
+}
+
+// Start initiates the net.Listener to listen for incoming connections
+func (proxy *proxyService) Start() error {
+	logger := proxy.logger
+
+	logger.Infof("Starting service")
+
+	if proxy.done {
+		return fmt.Errorf("cannot call Start on stopped ProxyService")
+	}
+
+	// TODO: why do we start a new goroutine here?
```

Comment:
The ProxyService `Start` method must be `non-blocking`. `http.Server` is blocking